src/storage/src/source/types.rs

```diff
@@ -83,22 +74,22 @@ pub trait SourceRender {
 /// Source-agnostic wrapper for messages. Each source must implement a
 /// conversion to Message.
 #[derive(Debug, Clone)]
-pub struct SourceMessage<Key, Value> {
+pub struct SourceMessage {
     /// The message key
-    pub key: Key,
+    pub key: Row,
```

Comment:
Is there any concern in forcing this additional allocation? I would have guessed this would become `Option<Row>`, but maybe I'm overestimating ingestions' memory pressure

Reply:
`Row` is non-allocating, like Vec, when it's empty so there isn't any additional memory pressure here

Reply:
TIL!